containers/src/Data/Sequence/Internal.hs

```diff
@@ -650,13 +650,42 @@ type Digit23 a = Node a
 -- If we used a higher-order nested type with MPTC, we could probably use a
 -- class, but as it is we have to build up @map23@ explicitly through the
 -- recursion.
+--
+-- Note: @aptyMiddle@ is polymorphic-recursive on @b@ and @c@, not @a@.
+--
+-- Invariants:
+--
+-- > 1. Viewing the various trees as the lists they represent:
+-- >
+-- >    toListFTN result
+-- >      =  (firstf <$> (toListThinN m ++ toListD sf)
+-- >      ++ (map23  <$> toListFTE fs <*> (toListD pr ++ toListThinN m ++ toListD sf))
+-- >      ++ (lastf  <$> (toListD pr ++ toListThinN m))
+-- >
+-- > 2. s = size m + size pr + size sf
+-- >
+-- > 3. size (firstf y) = size (lastf y) = size (map23 x y) = size y
+-- >      for any (x :: a) (y :: b)
+--
+-- Projecting invariant 1 on sizes, using 2 and 3 to simplify, we have the
+-- following corollary.
+-- It is weaker than invariant 1, but it may be easier to keep track of.
+--
+-- > 1a. size result = s * (size fs + 1) + size m
+--
+-- In invariant 1, the types of the auxiliary functions are as follows:
```

Comment:
Could you either move these auxiliary functions to the top or mention at the top that they are described below? When I first read this I immediately wondered if these were real functions in the module somewhere, or if I was supposed to guess their meanings from their names, or what.